### DISCUSSION TITLE
Validate a list of files in PowerShell with Pester

### DISCUSSION THREAD

**Discussion by erossini** [pester/Pester#2344] [Q&A] [answered]
I want to create a PowerShell script that can delete a bunch of file in given paths following a particular  match (in the following example all the files that contain _TMP_). This is the function:

    function CheckFilesToDelete([string[]]$fList) {
        [System.Collections.Generic.List[System.IO.DirectoryInfo]]$fileList

        foreach($folderName in $fList)
        {
	        $path = "$folderName"

	        ($list = @(Get-ChildItem "$path" -File -Recurse | 
                     Where-Object { $_.Name -match '.*TMP.*' })) > null
            if($list.Count -gt 0) { $fileList.AddRange($list) }
        }

        return $fileList
    }

Now, I want to create the test for this function. I mock the `Get-ChildItem` so I can verify what files the function has to return. I like to have all the details of the files (such as name, creation date and so on)

    Describe "Validate files to delete" {
        Context "validate files" {
            It "should return a list of expected files" {
                [string[]]$folderList = "Tests"
                $tmpPath = Join-Path $path $_

                [System.Collections.Generic.List[System.IO.DirectoryInfo]]
                $expected = @('1_TMP.txt','2_TMP.txt')

	            Mock Get-ChildItem {
	                [PSCustomObject]@{ Name = '1_TMP.txt' },
	                [PSCustomObject]@{ Name = 'SmokeTest.txt' },
	                [PSCustomObject]@{ Name = '2_TMP.txt' },
	                [PSCustomObject]@{ Name = 'qq02000.doc' }
	            }

                $list = CheckFilesToDelete -EnvPath $path 
                        -fList $folderList -ErrorAction Stop
            }
        }
    }

The problems I'm facing are:

- how can I display the `$list` to check the list of files
- how to compare `$list` with `$expected`

**Comment by fflaten** [comment]
Hi.

You can display `$list` by setting a breakpoint and debugging it, or adding `Write-Host $list` in the test temporary

If they should be equal, you can compare ex filenames using:
```powershell
$expected.Name | Sort-Object | Should -Be ($list.Name | Sort-Object)`

# sorting to avoid order-issue
```

**Tip:** For a simple filename-filter I'd recommend `$list = @(Get-ChildItem -Path "$path" -File -Recurse -Filter '*TMP*')`. It will process the filter in the filesystem-provider which is a lot faster than returning every file and filter in PowerShell like `Where-Object` and `Get-ChildItem -Include '*TMP*'` would.

For the future, it helps with a valid repro. This code won't run as-is due to wrong types in Mock (won't work with List.AddRange()`) an List-objects using `DirectoryInfo`-type when files are `FileInfo`.

**Reply by erossini** [reply]
Sorry, I didn't understand. How can I change the Mock?

**Reply by fflaten** [reply, marked as answer]
```powershell
BeforeAll {
    function CheckFilesToDelete([string[]]$fList) {
        $fileList = [System.Collections.Generic.List[System.IO.FileInfo]]::new()

        foreach ($folderName in $fList) {
            $path = "$folderName"

            # Using -Filter for improved performance for filename filter
            # Convert to typed array so .AddRange() doesn't complain about object[] parameter
            $list = [System.IO.FileInfo[]]@(Get-ChildItem -Path $path -File -Recurse -Filter '*TMP*')
            if ($list.Count -gt 0) { $fileList.AddRange($list) }
        }

        return $fileList
    }
}

Describe 'Validate files to delete' {
    Context 'validate files' {
        It 'should return a list of expected files' {
            [string[]]$folderList = 'Tests'

            $expected = [System.Collections.Generic.List[System.IO.FileInfo]]::new()
            $expected.Add([System.IO.FileInfo]::new('1_TMP.txt'))
            $expected.Add([System.IO.FileInfo]::new('2_TMP.txt'))

            Mock Get-ChildItem {
                $arr = @(
                    [System.IO.FileInfo]::new('1_TMP.txt'),
                    [System.IO.FileInfo]::new('SmokeTest.txt'),
                    [System.IO.FileInfo]::new('2_TMP.txt'),
                    [System.IO.FileInfo]::new('qq02000.doc'))

                # Mocking the -Filter parameter
                $arr | Where-Object Name -Like $Filter
            }

            $list = CheckFilesToDelete -EnvPath $path -fList $folderList -ErrorAction Stop

            # Asserting by name since we don't have consistent paths in mocked objects
            $list.Name | Sort-Object | Should -Be ($expected.Name | Sort-Object Name)
        }
    }
}
```

**Reply by erossini** [reply]
Thank you so much for this code!

**Comment by erossini** [comment]
Thank you!

**Comment by erossini** [comment]
Is it possible to check the `CreationTime`? How can I create a `$expected` list with the `CreationTime`?

If I try

```
$expected = [System.Collections.Generic.List[System.IO.FileInfo]]::new()
$f = [System.IO.FileInfo]::new('SmokeTest.txt')
$f.CreationTime = [DateTime]::ParseExact('2023-01-01 22:00', "yyyy-MM-dd HH:mm", $null) 
$expected.Add($f)
```

I got this error

> Validate files to delete.validate files with date.should return a list of expected files (mock) 4ms (3ms|1ms)
 FileNotFoundException: Could not find file 'C:\windows\system32\SmokeTest.txt'.
 SetValueInvocationException: Exception setting "CreationTime": "Could not find file 'C:\windows\system32\SmokeTest.txt'."

Thank you so much in advance.

**Reply by fflaten** [reply]
Modifying attributes in `FileInfo/DirectoryInfo`will try to trigger the update in the filesystem as well. You can use Pester's `New-MockObject` with `-Properties @{ .. }` to create the mock with a the specified properties set.

```powershell
It 'should return a list of expected files' {
  [string[]]$folderList = 'Tests'

  $expected = [System.Collections.Generic.List[System.IO.FileInfo]]::new()
  $expected.Add([System.IO.FileInfo]::new('1_TMP.txt'))
  $expected.Add([System.IO.FileInfo]::new('2_TMP.txt'))

  Mock Get-ChildItem {
      $arr = @(
          (New-MockObject -Type 'System.IO.FileInfo' -Properties @{ Name = '1_TMP.txt'; CreationTime = [datetime]'2020-01-01 21:00:00' }),
          (New-MockObject -Type 'System.IO.FileInfo' -Properties @{ Name = 'SmokeTest.txt'; CreationTime = [datetime]'2023-01-01 22:00:00' }),
          (New-MockObject -Type 'System.IO.FileInfo' -Properties @{ Name = '2_TMP.txt'; CreationTime = [datetime]'2022-01-01 22:00:00' }),
          (New-MockObject -Type 'System.IO.FileInfo' -Properties @{ Name = 'qq02000.doc'; CreationTime = [datetime]'2021-01-01 22:15:00' })
      )

      # Mocking the -Filter parameter if used
      if ($PesterBoundParameters.Filter) {
          return $arr | Where-Object Name -Like $PesterBoundParameters.Filter
      }

      return $arr
  }

  $list = CheckFilesToDelete -EnvPath $path -fList $folderList -ErrorAction Stop

  $threshold = [datetime]'2019-01-01 22:00:00'
  $actual = $list | Where-Object CreationTime -GT $threshold | Sort-Object Name
  $actual.Name | Should -Be ($expected.Name | Sort-Object Name)
}
```

**Reply by erossini** [reply]
Thank you so much!
